cmd/build/v2/hasher.go

```diff
@@ -0,0 +1,138 @@
+	projectCommitType      = "commit"
+)
+
+type repositoryCommitRetriever interface {
+	GetSHA() (string, error)
+	GetTreeHash(string) (string, error)
+}
+
+type serviceHasher struct {
+	gitRepoCtrl repositoryCommitRetriever
+
+	projectCommit string
+}
+
+func newServiceHasher(gitRepoCtrl repositoryCommitRetriever) *serviceHasher {
+	return &serviceHasher{
+		gitRepoCtrl: gitRepoCtrl,
+	}
+}
+
+// HashProjectCommit hashes the
+func (sh *serviceHasher) HashProjectCommit(buildInfo *model.BuildInfo) string {
+	if sh.projectCommit == "" {
+		var err error
+		sh.projectCommit, err = sh.gitRepoCtrl.GetSHA()
+		if err != nil {
+			oktetoLog.Infof("could not get repository sha: %w", err)
+		}
+	}
+	return sh.hash(buildInfo, projectCommitType, sh.projectCommit)
+}
+
+func (sh serviceHasher) HashBuildContext(buildInfo *model.BuildInfo) string {
+	buildContext := buildInfo.Context
+	if buildContext == "" {
+		buildContext = "."
+	}
+	treeHash, err := sh.gitRepoCtrl.GetTreeHash(buildContext)
+	if err != nil {
+		oktetoLog.Info("error trying to get tree hash for build context '%s': %w", buildContext, err)
+	}
+	return sh.hash(buildInfo, buildContextCommitType, treeHash)
+}
+
+func (sh serviceHasher) HashService(buildInfo *model.BuildInfo) string {
+	if LoadBoolean(OktetoSmartBuildUsingContextEnvVar) {
+		return sh.HashBuildContext(buildInfo)
+	}
+	return sh.HashProjectCommit(buildInfo)
+}
+
+func (sh serviceHasher) hash(buildInfo *model.BuildInfo, commitType, commitHash string) string {
+	args := []string{}
+	for _, arg := range buildInfo.Args {
+		args = append(args, arg.String())
+	}
+	argsText := strings.Join(args, ";")
+
+	secrets := []string{}
+	for key, value := range buildInfo.Secrets {
+		secrets = append(secrets, fmt.Sprintf("%s=%s", key, value))
+	}
+	secretsText := strings.Join(secrets, ";")
+
+	// We use a builder to avoid allocations when building the string
+	var b strings.Builder
+
+	fmt.Fprintf(&b, "%s:%s;", commitType, commitHash)
+	fmt.Fprintf(&b, "target:%s;", buildInfo.Target)
+	fmt.Fprintf(&b, "build_args:%s;", argsText)
+	fmt.Fprintf(&b, "secrets:%s;", secretsText)
+	fmt.Fprintf(&b, "context:%s;", buildInfo.Context)
+	fmt.Fprintf(&b, "dockerfile:%s;", buildInfo.Dockerfile)
+	if commitType == buildContextCommitType {
+		fmt.Fprintf(&b, "dockerfile_content:%s;", getDockerfileContent(buildInfo.Dockerfile))
+	}
+	fmt.Fprintf(&b, "image:%s;", buildInfo.Image)
+
+	oktetoBuildHash := sha256.Sum256([]byte(b.String()))
+	return hex.EncodeToString(oktetoBuildHash[:])
+}
+
+// getDockerfileContent returns the content of the Dockerfile
+func getDockerfileContent(dockerfilePath string) string {
+	content, err := os.ReadFile(dockerfilePath)
+	if err != nil {
+		oktetoLog.Info("error trying to read Dockerfile: %w", err)
+		return ""
+	}
+	encodedFile := sha256.Sum256(content)
+	return hex.EncodeToString(encodedFile[:])
+}
+
+// LoadBoolean loads a boolean environment variable and returns it value
+func LoadBoolean(k string) bool {
```

Comment:
I feel we already have this logic. Is it?

Reply:
Yup https://github.com/okteto/okteto/pull/4051#discussion_r1399350133